File: src/modules/decision-twin/scenario_simulation_agent.ts
import { ToolDecorator as Tool, ExecutionContext, z } from '@nitrostack/core';

export interface ScenarioSimulationParams {
  vibrationLevel?: number;
  temperature?: number;
  hourlyDowntimeCost?: number;
  activeOrders?: number;
  delayDays?: number;
  capacityPct?: number;
}

export interface StrategyDetail {
  id: string;
  name: string;
  badge?: string;
  badgeColor?: string;
  totalExpectedCost: number;
  repairCost: number;
  downtimeLoss: number;
  downtimeHours: number;
  failureRiskPct: number;
  deliveryRiskPct: number;
  resilienceScore: number;
  recommendationRank: number;
  pros: string[];
  cons: string[];
}

export interface ScenarioSimulationResult {
  parametersEvaluated: {
    vibrationLevel: number;
    temperature: number;
    hourlyDowntimeCost: number;
    activeOrders: number;
    delayDays: number;
    capacityPct: number;
    severityFactor: number;
  };
  bestStrategyId: string;
  strategies: Record<string, StrategyDetail>;
  costSavingsVsDelay: number;
}

/**
 * Calculate quantitative metrics for all 3 maintenance strategies.
 */
export function calculateStrategyMetrics(params: ScenarioSimulationParams = {}): ScenarioSimulationResult {
  const vibrationLevel = params.vibrationLevel ?? 6.5;
  const temperature = params.temperature ?? 85.0;
  const hourlyDowntimeCost = params.hourlyDowntimeCost ?? 12500.0;
  const activeOrders = params.activeOrders ?? 3;
  const delayDays = params.delayDays ?? 7;
  const capacityPct = params.capacityPct ?? 60.0;

  // Base severity coefficient
  const severityFactor = Math.max(1.0, (vibrationLevel / 5.0) * (temperature / 80.0));

  // --- 1. REPAIR NOW ---
  const rnDowntimeHrs = 4.0;
  const rnRepairCost = 45000.0;
  const rnDowntimeLoss = rnDowntimeHrs * hourlyDowntimeCost;
  const rnTotalCost = rnRepairCost + rnDowntimeLoss;
  const rnFailureRisk = 2.0;
  const rnDeliveryRisk = 5.0;
  const rnScore = severityFactor > 1.2 ? 92 : 80;

  // --- 2. DELAY REPAIR ---
  const baseDailyRisk = Math.min(0.25, 0.05 * severityFactor);
  const drFailureProb = Math.min(0.95, 1.0 - Math.exp(-baseDailyRisk * delayDays));
  const drCatastrophicDowntimeHrs = 48.0;
  const drCatastrophicRepairCost = 140000.0;
  const drDowntimeLoss = drCatastrophicDowntimeHrs * hourlyDowntimeCost;
  const drOrderPenalty = activeOrders * 25000.0;

  const drExpectedCost = drFailureProb * (drCatastrophicRepairCost + drDowntimeLoss + drOrderPenalty);
  const drFailureRisk = Number((drFailureProb * 100).toFixed(1));
  const drDeliveryRisk = Math.min(95.0, Number((drFailureProb * 90).toFixed(1)));
  const drScore = Math.max(10, 100 - Math.round(drFailureRisk * 1.1));

  // --- 3. REDUCED CAPACITY ---
  const deratedStress = severityFactor * Math.pow(capacityPct / 100.0, 2);
  const rcFailureProb = Math.min(0.35, 0.03 * deratedStress * delayDays);
  const rcCapacityLossPerDay = ((100.0 - capacityPct) / 100.0) * (hourlyDowntimeCost * 12);
  const rcRevenueLoss = rcCapacityLossPerDay * delayDays;
  const rcScheduledRepairCost = 38000.0;
  const rcPlannedDowntimeLoss = 4.0 * hourlyDowntimeCost * 0.7;

  const rcExpectedCost = rcRevenueLoss + rcScheduledRepairCost + rcPlannedDowntimeLoss + (rcFailureProb * drCatastrophicRepairCost);
  const rcFailureRisk = Number((rcFailureProb * 100).toFixed(1));
  const rcDeliveryRisk = Number(Math.min(60.0, (100.0 - capacityPct) * 0.75).toFixed(1));
  const rcScore = severityFactor <= 1.5 ? 85 : 72;

  const strategies: Record<string, StrategyDetail> = {
    repair_now: {
      id: 'repair_now',
      name: 'Repair Now (Immediate Maintenance)',
      badge: 'Immediate Action',
      badgeColor: 'cyan',
      totalExpectedCost: Math.round(rnTotalCost),
      repairCost: Math.round(rnRepairCost),
      downtimeLoss: Math.round(rnDowntimeLoss),
      downtimeHours: rnDowntimeHrs,
      failureRiskPct: rnFailureRisk,
      deliveryRiskPct: rnDeliveryRisk,
      resilienceScore: rnScore,
      recommendationRank: rnScore >= Math.max(drScore, rcScore) ? 1 : 2,
      pros: ['Eliminates catastrophic breakdown risk', 'Uses in-stock parts immediately', 'Fastest restoration to 100% capacity'],
      cons: ['Immediate 4-hour production stoppage'],
    },
    reduced_capacity: {
      id: 'reduced_capacity',
      name: `Operate at Reduced Capacity (De-rate to ${Math.round(capacityPct)}%)`,
      badge: 'Balanced De-Rate',
      badgeColor: 'amber',
      totalExpectedCost: Math.round(rcExpectedCost),
      repairCost: Math.round(rcScheduledRepairCost),
      downtimeLoss: Math.round(rcRevenueLoss + rcPlannedDowntimeLoss),
      downtimeHours: 4.0,
      failureRiskPct: rcFailureRisk,
      deliveryRiskPct: rcDeliveryRisk,
      resilienceScore: rcScore,
      recommendationRank: rnScore >= Math.max(drScore, rcScore) ? 2 : 1,
      pros: ['Maintains partial order throughput', 'Schedules maintenance during off-peak hours', 'Significant stress reduction'],
      cons: ['Accumulates daily capacity revenue loss', 'Requires active order rerouting'],
    },
    delay_repair: {
      id: 'delay_repair',
      name: `Delay Repair (${delayDays} Days)`,
      badge: 'High Risk Deferral',
      badgeColor: 'rose',
      totalExpectedCost: Math.round(drExpectedCost),
      repairCost: Math.round(drCatastrophicRepairCost * drFailureProb),
      downtimeLoss: Math.round(drDowntimeLoss * drFailureProb),
      downtimeHours: Number((drCatastrophicDowntimeHrs * drFailureProb).toFixed(1)),
      failureRiskPct: drFailureRisk,
      deliveryRiskPct: drDeliveryRisk,
      resilienceScore: drScore,
      recommendationRank: 3,
      pros: ['Zero immediate downtime today'],
      cons: ['Exponentially escalating failure risk', 'Potential 48h catastrophic outage', 'Severe financial penalty'],
    },
  };

  const bestId = Object.keys(strategies).reduce((a, b) =>
    strategies[a].resilienceScore > strategies[b].resilienceScore ? a : b
  );

  return {
    parametersEvaluated: {
      vibrationLevel,
      temperature,
      hourlyDowntimeCost,
      activeOrders,
      delayDays,
      capacityPct,
      severityFactor: Number(severityFactor.toFixed(2)),
    },
    bestStrategyId: bestId,
    strategies,
    costSavingsVsDelay: Math.round(drExpectedCost - strategies[bestId].totalExpectedCost),
  };
}

/**
 * ScenarioSimulationAgent MCP Controller & Tool Provider
 */
export class ScenarioSimulationAgent {
  /**
   * Tool 1: simulate_scenarios
   * Evaluates Repair Now, Delay Repair, and Reduced Capacity operational strategies.
   */
  @Tool({
    name: 'simulate_scenarios',
    description: 'Performs quantitative rule-based simulation comparing three operational strategies: Repair Now, Delay Repair, and Reduced Capacity.',
    inputSchema: z.object({
      vibrationLevel: z.number().optional().default(6.5).describe('Vibration level reading (mm/s)'),
      temperature: z.number().optional().default(85.0).describe('Operating temperature in Celsius'),
      hourlyDowntimeCost: z.number().optional().default(12500.0).describe('Hourly downtime cost in USD'),
      activeOrders: z.number().optional().default(3).describe('Active production orders affected'),
      delayDays: z.number().optional().default(7).describe('Deferred repair delay in days'),
      capacityPct: z.number().optional().default(60.0).describe('De-rated capacity percentage')
    })
  })
  async simulateScenarios(
    input: ScenarioSimulationParams,
    ctx?: ExecutionContext
  ): Promise<ScenarioSimulationResult> {
    if (ctx?.logger) {
      ctx.logger.info('Executing Scenario Simulation Agent strategy matrix', { ...input });
    }
    return calculateStrategyMetrics(input);
  }

  /**
   * Tool 2: evaluate_scenario_agent
   * Evaluates state event parameters and produces a blackboard proposal.
   */
  @Tool({
    name: 'evaluate_scenario_agent',
    description: 'Evaluates state event parameters for Decision Twin multi-agent workflow and returns agent blackboard state and proposal.',
    inputSchema: z.object({
      vibrationLevel: z.number().optional().default(6.5),
      temperature: z.number().optional().default(85.0),
      hourlyDowntimeCost: z.number().optional().default(12500.0),
      activeOrders: z.number().optional().default(3),
      delayDays: z.number().optional().default(7),
      capacityPct: z.number().optional().default(60.0)
    })
  })
  async evaluateScenarioAgent(input: ScenarioSimulationParams, ctx?: ExecutionContext) {
    if (ctx?.logger) {
      ctx.logger.info('Evaluating scenario simulation agent state proposal', { ...input });
    }

    const simulationResults = calculateStrategyMetrics(input);
    const bestStrat = simulationResults.strategies[simulationResults.bestStrategyId];

    const proposal = {
      source: 'scenario_simulation_agent',
      action: simulationResults.bestStrategyId,
      reason: `Numeric simulation ranks ${bestStrat.name} #1 (Score: ${bestStrat.resilienceScore}/100). Saves $${simulationResults.costSavingsVsDelay.toLocaleString()} compared to delaying repair.`,
      confidence: 0.88,
      matrix: simulationResults,
    };

    const trace = [
      `     [ScenarioSimulationAgent] Evaluated 3 strategies:`,
      `       1. Repair Now        : Expected Cost = $${simulationResults.strategies.repair_now.totalExpectedCost.toLocaleString()} (Risk: ${simulationResults.strategies.repair_now.failureRiskPct}%)`,
      `       2. Reduced Capacity  : Expected Cost = $${simulationResults.strategies.reduced_capacity.totalExpectedCost.toLocaleString()} (Risk: ${simulationResults.strategies.reduced_capacity.failureRiskPct}%)`,
      `       3. Delay Repair (${simulationResults.parametersEvaluated.delayDays}d): Expected Cost = $${simulationResults.strategies.delay_repair.totalExpectedCost.toLocaleString()} (Risk: ${simulationResults.strategies.delay_repair.failureRiskPct}%)`,
      `       Recommendation: ${simulationResults.bestStrategyId} (Savings vs Delay: $${simulationResults.costSavingsVsDelay.toLocaleString()})`,
    ];

    return {
      blackboard: { scenario_simulation_agent: simulationResults },
      agentsCompleted: ['scenario_simulation_agent'],
      proposals: [proposal],
      trace,
    };
  }
}

File: src/orchestrator/scenario_simulation_agent.py
"""ScenarioSimulationAgent for Decision Twin System.

Performs quantitative rule-based simulation comparing three operational strategies:
1. Repair Now (immediate_repair)
2. Delay Repair (delay_repair)
3. Reduced Capacity (reduced_capacity)

Computes expected costs, risk scores, downtime, delivery impact, and ROI matrices.
"""

from __future__ import annotations
import math
from typing import Dict, Any, List

def calculate_strategy_metrics(
    vibration_level: float = 6.5,
    temperature: float = 85.0,
    hourly_downtime_cost: float = 12500.0,
    active_orders: int = 3,
    delay_days: int = 7,
    capacity_pct: float = 60.0
) -> Dict[str, Any]:
    """Calculate quantitative metrics for all 3 maintenance strategies."""

    # Base severity coefficient
    severity_factor = max(1.0, (vibration_level / 5.0) * (temperature / 80.0))

    # --- 1. REPAIR NOW ---
    rn_downtime_hrs = 4.0
    rn_repair_cost = 45000.0
    rn_downtime_loss = rn_downtime_hrs * hourly_downtime_cost
    rn_total_cost = rn_repair_cost + rn_downtime_loss
    rn_failure_risk = 2.0  # % after repair
    rn_delivery_risk = 5.0  # % (reroute absorbs 4 hrs)
    rn_score = 92 if severity_factor > 1.2 else 80

    # --- 2. DELAY REPAIR ---
    # Failure probability escalates exponentially over delay days
    base_daily_risk = min(0.25, 0.05 * severity_factor)
    dr_failure_prob = min(0.95, 1.0 - math.exp(-base_daily_risk * delay_days))
    dr_catastrophic_downtime_hrs = 48.0
    dr_catastrophic_repair_cost = 140000.0  # secondary damage to motor/spindle
    dr_downtime_loss = dr_catastrophic_downtime_hrs * hourly_downtime_cost
    dr_order_penalty = active_orders * 25000.0

    dr_expected_cost = (
        dr_failure_prob * (dr_catastrophic_repair_cost + dr_downtime_loss + dr_order_penalty)
    )
    dr_failure_risk = round(dr_failure_prob * 100, 1)
    dr_delivery_risk = min(95.0, round(dr_failure_prob * 90, 1))
    dr_score = max(10, 100 - int(dr_failure_risk * 1.1))

    # --- 3. REDUCED CAPACITY ---
    # Capacity reduction lowers stress factor
    derated_stress = severity_factor * (capacity_pct / 100.0) ** 2
    rc_failure_prob = min(0.35, 0.03 * derated_stress * delay_days)
    rc_capacity_loss_per_day = (100.0 - capacity_pct) / 100.0 * (hourly_downtime_cost * 12)  # 12 operating hrs/day
    rc_revenue_loss = rc_capacity_loss_per_day * delay_days
    rc_scheduled_repair_cost = 38000.0  # planned off-peak repair cost
    rc_planned_downtime_loss = 4.0 * hourly_downtime_cost * 0.7  # 30% off-peak discount

    rc_expected_cost = rc_revenue_loss + rc_scheduled_repair_cost + rc_planned_downtime_loss + (rc_failure_prob * dr_catastrophic_repair_cost)
    rc_failure_risk = round(rc_failure_prob * 100, 1)
    rc_delivery_risk = round(min(60.0, (100.0 - capacity_pct) * 0.75), 1)
    rc_score = 85 if severity_factor <= 1.5 else 72

    strategies = {
        "repair_now": {
            "id": "repair_now",
            "name": "Repair Now (Immediate Maintenance)",
            "total_expected_cost": round(rn_total_cost, 2),
            "repair_cost": round(rn_repair_cost, 2),
            "downtime_loss": round(rn_downtime_loss, 2),
            "downtime_hours": rn_downtime_hrs,
            "failure_risk_pct": rn_failure_risk,
            "delivery_risk_pct": rn_delivery_risk,
            "resilience_score": rn_score,
            "recommendation_rank": 1 if rn_score >= max(dr_score, rc_score) else 2,
            "pros": ["Eliminates catastrophic breakdown risk", "Uses in-stock parts immediately", "Fastest restoration to 100% capacity"],
            "cons": ["Immediate 4-hour production stoppage"],
        },
        "reduced_capacity": {
            "id": "reduced_capacity",
            "name": "Operate at Reduced Capacity (De-rate to " + str(int(capacity_pct)) + "%)",
            "total_expected_cost": round(rc_expected_cost, 2),
            "repair_cost": round(rc_scheduled_repair_cost, 2),
            "downtime_loss": round(rc_revenue_loss + rc_planned_downtime_loss, 2),
            "downtime_hours": 4.0,  # Scheduled off-peak
            "failure_risk_pct": rc_failure_risk,
            "delivery_risk_pct": rc_delivery_risk,
            "resilience_score": rc_score,
            "recommendation_rank": 2 if rn_score >= max(dr_score, rc_score) else 1,
            "pros": ["Maintains partial order throughput", "Schedules maintenance during off-peak hours", "Significant stress reduction"],
            "cons": ["Accumulates daily capacity revenue loss", "Requires active rerouting"],
        },
        "delay_repair": {
            "id": "delay_repair",
            "name": "Delay Repair (" + str(delay_days) + " Days)",
            "total_expected_cost": round(dr_expected_cost, 2),
            "repair_cost": round(dr_catastrophic_repair_cost * dr_failure_prob, 2),
            "downtime_loss": round(dr_downtime_loss * dr_failure_prob, 2),
            "downtime_hours": round(dr_catastrophic_downtime_hrs * dr_failure_prob, 1),
            "failure_risk_pct": dr_failure_risk,
            "delivery_risk_pct": dr_delivery_risk,
            "resilience_score": dr_score,
            "recommendation_rank": 3,
            "pros": ["Zero immediate downtime today"],
            "cons": ["Exponentially escalating failure risk", "Potential 48h catastrophic outage", "Severe financial penalty"],
        },
    }

    # Best recommendation determination
    best_id = max(strategies.keys(), key=lambda k: strategies[k]["resilience_score"])
    
    return {
        "parameters_evaluated": {
            "vibration_level": vibration_level,
            "temperature": temperature,
            "hourly_downtime_cost": hourly_downtime_cost,
            "active_orders": active_orders,
            "delay_days": delay_days,
            "capacity_pct": capacity_pct,
            "severity_factor": round(severity_factor, 2),
        },
        "best_strategy_id": best_id,
        "strategies": strategies,
        "cost_savings_vs_delay": round(dr_expected_cost - strategies[best_id]["total_expected_cost"], 2),
    }


def scenario_simulation_agent(state: Dict[str, Any]) -> Dict[str, Any]:
    """LangGraph agent node for Scenario Simulation.

    Takes DecisionTwinState, evaluates numerical options, and updates blackboard & proposals.
    """
    event = state.get("event", {})
    bb = state.get("blackboard", {})
    sensor_data = bb.get("sensor_agent", {}).get("readings", {})
    finance_data = bb.get("finance_agent", {})

    vibration = float(event.get("vibration_level") or sensor_data.get("vibration", 6.5))
    temperature = float(event.get("temperature") or sensor_data.get("temperature", 85.0))
    hourly_cost = float(finance_data.get("hourly_downtime_cost", 12500.0))

    simulation_results = calculate_strategy_metrics(
        vibration_level=vibration,
        temperature=temperature,
        hourly_downtime_cost=hourly_cost,
    )

    best_strat = simulation_results["strategies"][simulation_results["best_strategy_id"]]

    proposal = {
        "source": "scenario_simulation_agent",
        "action": simulation_results["best_strategy_id"],
        "reason": (
            f"Numeric simulation ranks {best_strat['name']} #1 (Score: {best_strat['resilience_score']}/100). "
            f"Saves ${simulation_results['cost_savings_vs_delay']:,} compared to delaying repair."
        ),
        "confidence": 0.88,
        "matrix": simulation_results,
    }

    return {
        "blackboard": {"scenario_simulation_agent": simulation_results},
        "agents_completed": ["scenario_simulation_agent"],
        "proposals": [proposal],
        "trace": [
            f"     [ScenarioSimulationAgent] Evaluated 3 strategies:",
            f"       1. Repair Now        : Expected Cost = ${simulation_results['strategies']['repair_now']['total_expected_cost']:,} (Risk: {simulation_results['strategies']['repair_now']['failure_risk_pct']}%)",
            f"       2. Reduced Capacity  : Expected Cost = ${simulation_results['strategies']['reduced_capacity']['total_expected_cost']:,} (Risk: {simulation_results['strategies']['reduced_capacity']['failure_risk_pct']}%)",
            f"       3. Delay Repair ({simulation_results['parameters_evaluated']['delay_days']}d): Expected Cost = ${simulation_results['strategies']['delay_repair']['total_expected_cost']:,} (Risk: {simulation_results['strategies']['delay_repair']['failure_risk_pct']}%)",
            f"       Recommendation: {simulation_results['best_strategy_id']} (Savings vs Delay: ${simulation_results['cost_savings_vs_delay']:,})",
        ],
    }
